delfin/drivers/dell_emc/vnx_block/navi_handler.py

```diff
@@ -0,0 +1,429 @@
+# Copyright 2020 The SODA Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import copy
+import re
+import time
+
+import six
+from oslo_log import log as logging
+from oslo_utils import units
+
+from delfin import cryptor
+from delfin import exception
+from delfin.drivers.utils.navicli_client import NaviClient
+from delfin.drivers.utils.tools import Tools
+
+LOG = logging.getLogger(__name__)
+
+
+class NaviHandler(object):
+    """Common class for EMC VNX storage system."""
+
+    USER_SECURITY_API = 'naviseccli -AddUserSecurity -User %(username)s ' \
+                        '-password %(password)s -h %(host)s -scope 0 ' \
+                        '-t %(timeout)d'
+    REMOVEUSERSECURITY_API = 'naviseccli -h %(host)s -RemoveUserSecurity'
+    GET_AGENT_API = 'naviseccli -h %(host)s getagent'
+    GET_DOMAIN_API = 'naviseccli -h %(host)s domain -list'
+    GET_STORAGEPOOL_API = 'naviseccli -h %(host)s storagepool -list'
+    GET_RAIDGROUP_API = 'naviseccli -h %(host)s getall -rg'
+    GET_DISK_API = 'naviseccli -h %(host)s getdisk'
+    GET_LUN_API = 'naviseccli -h %(host)s lun -list'
+    GET_GETALLLUN_API = 'naviseccli -h %(host)s getall -lun'
+    GET_LOG_API = 'naviseccli -h %(host)s getlog ' \
+                  '-date %(begin_time)s %(end_time)s'
+
+    TIME_PATTERN = '%m/%d/%Y %H:%M:%S'
+    DATE_PATTERN = '%m/%d/%Y'
+    ONE_DAY_SCE = 24 * 60 * 60
+    SOCKET_TIMEOUT = 5
+
+    def __init__(self, **kwargs):
+        ssh_access = kwargs.get('ssh')
+        if ssh_access is None:
+            raise exception.InvalidInput('Input navicli_access is missing')
+        self.navi_host = ssh_access.get('host')
+        self.navi_port = ssh_access.get('port')
+        self.navi_username = ssh_access.get('username')
+        self.navi_password = cryptor.decode(ssh_access.get('password'))
+        self.navi_timeout = ssh_access.get('conn_timeout')
+        if self.navi_timeout is None:
+            self.navi_timeout = NaviClient.SOCKET_TIMEOUT
+
+    def login(self, host_ip=None):
```

Comment:
Which registration interface will be used for navi_cli login? Do we need a new interface to support navicli?  if it is ssh , is public_key required! as of now pub_key is a required para for ssh..

Reply:
At present, vnx bloc system only supports navicli access, but does not support traditional SSH mode and public key. At present, the system needs a new interface to support navicli

Reply:
Vnx bloc system only supports navicli access, but does not support traditional SSH mode and public key. The system needs a new or modified interface to support navicli

Reply:
How do we test this driver now without navicli interface from delfin framework ? @ThisIsClark @wisererik We need to support new registration interface for this.. What is your suggestion ?

Reply:
Before testing on the DME, the DME still uses the SSH access interface, but it does not need to load the public key

Reply:
Before testing on the system, the system still uses the SSH access interface, but it does not need to load the public key